hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,749 @@
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, err
+}
+
+func (gr *GenericReconciler) getStatus(ctx context.Context, id string, data *ReconcileMetadata) (genruntime.ArmTransformer, error) {
+	_, typedArmSpec, err := ResourceSpecToArmResourceSpec(gr, data.metaObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: do we tolerate not exists here?
+	armStatus, err := NewEmptyArmResourceStatus(data.metaObj)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to construct empty ARM status object for resource with ID: %q", id)
+	}
+
+	// Get the resource
+	err = gr.ARMClient.GetResource(ctx, id, typedArmSpec.GetApiVersion(), armStatus)
+	if data.log.V(0).Enabled() { // TODO: CHange this level
+		statusBytes, err := json.Marshal(armStatus)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to serialize ARM status to JSON for debugging")
+		}
+		data.log.V(0).Info("Got ARM status", "status", string(statusBytes))
+	}
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get resource with ID: %q", id)
+	}
+
+	// Convert the ARM shape to the Kube shape
+	status, err := NewEmptyStatus(data.metaObj)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to construct empty status object for resource with ID: %q", id)
+	}
+
+	// TODO: need to use KnownOwner?
+	owner := data.metaObj.Owner()
+	correctOwner := genruntime.KnownResourceReference{
+		Name: owner.Name,
+	}
+
+	// Fill the kube status with the results from the arm status
+	err = status.PopulateFromArm(correctOwner, ValueOfPtr(armStatus)) // TODO: PopulateFromArm expects a value... ick
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't convert ARM status to Kubernetes status")
+	}
+
+	return status, nil
+}
+
+func (gr *GenericReconciler) updateMetaObject(
+	metaObj genruntime.MetaObject,
+	deployment *armclient.Deployment,
+	status genruntime.ArmTransformer) error {
+
+	controllerutil.AddFinalizer(metaObj, GenericControllerFinalizer)
+
+	sig, err := SpecSignature(metaObj)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute resource spec hash")
+	}
+
+	annotations := make(map[string]string)
+	annotations[ResourceSigAnnotationKey] = sig
+	annotations[DeploymentIdAnnotation] = deployment.Id
+	annotations[DeploymentNameAnnotation] = deployment.Name
+
+	// TODO: Do we want to just use Azure's annotations here? I bet we don't? We probably want to map
+	// TODO: them onto something more robust? For now just use Azure's though.
+	annotations[ResourceStateAnnotation] = string(deployment.Properties.ProvisioningState)
+
+	if deployment.IsTerminalProvisioningState() {
+		if deployment.Properties.ProvisioningState == armclient.FailedProvisioningState {
+			annotations[ResourceErrorAnnotation] = deployment.Properties.Error.String()
+		} else if len(deployment.Properties.OutputResources) == 0 {
+			return errors.Errorf("Template deployment didn't have any output resources")
```

Comment:
Is there a way to structure this logic that doesn't involved returning from the middle block?

Reply:
We could swap the second and third blocks, if it's better to return in the last? I've done that for now but will leave this unresolved.

Reply:
Looks good. Having a return in the middle block was really hard to read.